pwiz_tools/Skyline/SkylineNightly/Nightly.cs

```diff
@@ -412,24 +410,48 @@ public string Run()
                     return result;
                 }
                 Log("SkylineTester started");
-                if (!skylineTesterProcess.WaitForExit(durationSeconds * 1000))
-                {
-                    SaveErrorScreenshot();
-                    Log(result = "SkylineTester has exceeded its " + durationSeconds +
-                                 " second WaitForExit timeout.  You should investigate.");
-                }
-                else if (skylineTesterProcess.ExitCode == 0xDEAD)
-                {
-                    // User killed, don't post
-                    Log(result = SkylineTesterStoppedByUser);
-                    return result;
-                }
-                else
+
+                var endTime = skylineTesterProcess.StartTime.AddHours(durationHours);
+                var originalEndTime = endTime;
+                for (;; Thread.Sleep(1000))
                 {
-                    Log("SkylineTester finished");
+                    if (skylineTesterProcess.HasExited)
+                    {
+                        if (skylineTesterProcess.ExitCode == 0xDEAD)
+                        {
+                            // User killed, don't post
+                            Log(result = SkylineTesterStoppedByUser);
+                            return result;
+                        }
+                        Log("SkylineTester finished");
+                        break;
+                    }
+                    else if (DateTime.Now > endTime.AddMinutes(30)) // 30 minutes grace before we kill SkylineTester
+                    {
+                        SaveErrorScreenshot();
+                        Log(result = "SkylineTester has exceeded its " + durationHours + " hour runtime.  You should investigate.");
+                        break;
+                    }
+
+                    if (endTime == originalEndTime)
+                    {
+                        if (logMonitor.IsHanging())
+                        {
+                            // extend the end time until 12pm
```

Comment:
This is presumably so somebody can get there in time to attach a debugger? If so, maybe add that to the comment.